File: app/src/main/java/wannabit/io/cosmostaion/dao/FeeInfo.java
package wannabit.io.cosmostaion.dao;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class FeeInfo {
    public String title = "";
    public String msg = "";
    public ArrayList<FeeData> feeDatas = new ArrayList<>();

    public FeeInfo(String data) {
        if (data == null) return;
        for (String rawData : data.split(",")) {
            feeDatas.add(new FeeData(rawData));
        }
    }

    public class FeeData {
        public String denom;
        public BigDecimal gasRate;

        public FeeData(String data) {
            Pattern p = Pattern.compile("([0-9])*\\.?[0-9]*");
            Matcher m = p.matcher(data);
            if (m.find()) {
                String rawGasRate = m.group();
                String denom = data.substring(m.end());
                this.denom = denom;
                this.gasRate = new BigDecimal(rawGasRate);
            }
        }
    }
}

File: app/src/main/java/wannabit/io/cosmostaion/data/model/res/FeeInfo.kt
package wannabit.io.cosmostaion.data.model.res

import android.os.Parcelable
import kotlinx.parcelize.Parcelize
import java.math.BigDecimal

data class FeeInfo(
    var title: String? = "",
    var msg: String? = "",
    var feeDatas: List<FeeData> = mutableListOf()
) {
    constructor(data: String?) : this() {
        if (data == null) return
        feeDatas = data.split(",").map { FeeData(it) }
    }
}

@Parcelize
data class FeeData(
    var denom: String? = null,
    var gasRate: BigDecimal? = null
) : Parcelable {
    constructor(data: String?) : this() {
        val regex = Regex("[0-9]*\\.?[0-9]*")
        val matchResult = regex.find(data ?: "")
        if (matchResult != null) {
            val rawGasRate = matchResult.value
            val denomIndex = rawGasRate.length

            denom = data?.substring(denomIndex)
            gasRate = BigDecimal(rawGasRate)
        }
    }
}